### PR TITLE
ensure iopub subscriptions propagate prior to accepting websocket connections

### PR DESCRIPTION
**Pull request by SylvainCorlay**
Nudge kernel with info request upon opening websocket until some IOPub message is received.

- repeat info requests every 500ms until iopub is received
- busy kernels are not nudged, which would block connections to busy kernels indefinitely

cf jupyter/jupyter_client#593

### CODE REVIEW

**Comment by jasongrout** [comment]
I thought that there was one set of zmq channels (shell, iopub, etc.) shared across all websocket connections (but I haven't looked at the code carefully recently). Is that wrong, or is that changing in this PR?

**Comment by SylvainCorlay** [comment]
> I thought that there was one set of zmq channels (shell, iopub, etc.) shared across all websocket connections (but I haven't looked at the code carefully recently). Is that wrong, or is that changing in this PR?

I think that it is a correct statement.

**Comment by minrk** [comment]
> I thought that there was one set of zmq channels

Each websocket client gets its own zmq connection to the kernel, created [here](https://github.com/jupyter/notebook/blob/e946154112daa5ef997d7f549953cce91b336d3d/notebook/services/kernels/handlers.py#L123-L129) using `KernelManager.connect_iopub()`, etc.

This is what allows the notebook server to not need client-identifying logic, since reply-routing on e.g. stdin, stream channels is handled in the kernel. IOPub messages are indeed wastefully duplicated, however.

File: notebook/services/kernels/handlers.py
Comment (minrk) [inline notebook/services/kernels/handlers.py:180]: If this is getting retried more than once or twice, it may be worth logging a more informative message.

Maybe a counter and log with debug if count <= 1, and warning with the count itself if count > 1?

File: notebook/services/kernels/handlers.py
Comment (SylvainCorlay) [reply notebook/services/kernels/handlers.py:326]: > I don't think nudge on reconnect is appropriate. 

This is the branch that is used when restarting a kernel, and which causes the issue with "restart and run all", so I think we should nudge on reconnect...

(The trivial `connected` Future is what I had originally).

File: notebook/services/kernels/handlers.py
Comment (SylvainCorlay) [reply notebook/services/kernels/handlers.py:315]: > This method is not being turned into a coroutine. 

Thanks! This is a leftover from earlier versions of the PR. Making it *not* a coroutine now.

File: notebook/services/kernels/handlers.py
Comment (SylvainCorlay) [reply notebook/services/kernels/handlers.py]: I have added a counter for the calls to nudge, and am now logging the nudge attempts.

File: notebook/services/kernels/handlers.py
Comment (minrk) [inline notebook/services/kernels/handlers.py:182]: If this is taking a while, it's probably worth logging at a higher level:

```suggestion
                log = self.log.warning if nudge_count % 10 == 0 else self.log.debug
                log("Nudging attempt %s on kernel %s" % (nudge_count, self.kernel_id))
```

File: notebook/services/kernels/handlers.py
Comment (minrk) [inline notebook/services/kernels/handlers.py:180]:
```suggestion
            nonlocal nudge_count
            nudge_count += 1
```

Comment (SylvainCorlay) [reply notebook/services/kernels/handlers.py:180]: I am not sure about this since we are passing cout to the callback. You want to also remove the count parameter?

Comment (minrk) [reply notebook/services/kernels/handlers.py:180]: Oops! No, I didn't notice the closure. Don't mind me!

**Comment by minrk** [comment]
Having poked and prodded quite a bit now, I understand how both branches need a nudge in case of restart:

- zmq sockets are not recreated when the kernel is restarted (they shouldn't need to be, as zmq does its own underlying auto reconnect).
- There is no similar nudge operation triggered on restart for 'existing' connections
- To prod connections, clients tend to recreate websocket connection on restart (this is not strictly required, but it is a standard practice and may be needed in some cases where the zmq ports change)
- a 'reconnecting' websocket may not actually mean a new zmq socket (the 'restoring connection' branch is taken), if the same session id is re-used

So it's possible for a 'new' websocket connection, as part of the client's restart steps, to re-use zmq sockets and thus never nudging the new kernel's sockets.

The downside of doing this nudge on every ws connect, even on established, already-nudged zmq sockets, is that websocket connections will not be accepted while the kernel is busy, even though they should be. Currently, it is valid to connect and start sending requests without waiting for iopub, but this can result in the issues folks are seeing. The most robust and efficient nudge would be on:

1. new zmq connections, and
2. kernel restarts for all existing sockets

where requests on *existing, open sockets* are blocked until the nudge resolves, not just connection-opening. That's more complicated, and I think we can go with this solution as first pass, and refine it later, especially if the delay is bothering folks.

The last thing I want to look at before merging is what happens if nudge is still outstanding when the connections are closed, then I think this is okay to merge. Thanks @SylvainCorlay!

**Comment by minrk** [comment]
I noticed that there are test failures caused by a change in status messages delivered. The nudge waits for any iopub message, then resolves, which means it will tend to resolve on the 'busy' status message, and the 'idle' status message will end up propagating to the client after resolving the connection. It may be better to wait for specifically the idle message associated with the info request to preserve message expectations, even though this is not strictly required.

**Comment by SylvainCorlay** [comment]
@minrk I think this is ready to go.

When this gets in I will update the `jupyter_server` PR.

**Review by minrk — APPROVED** [review]

**Comment by Zsailer** [comment]
Thanks, @SylvainCorlay, @minrk, and @jasongrout!

File: notebook/services/kernels/handlers.py
Comment (jasongrout) [inline notebook/services/kernels/handlers.py:170]: Just checking - the function argument is f and the loop variable is f. Is that going to cause a problem?

**Comment by jasongrout** [comment]
@kevin-bates - what is the plan for a release with this PR merged? We're seeing consistent problems that this PR is designed to fix, and would love to have a release fixing the problem. I'm happy to help with the release process if that is helpful.

**Comment by Zsailer** [comment]
I can work on a patch release, since @kevin-bates is on vacation this month.

**Comment by jasongrout** [comment]
Awesome, thanks @Zsailer! Again, I'm happy to help if needed.

**Comment by Zsailer** [comment]
Oops, I thought I was a maintainer on PyPI for this repo, but it looks like I was mistaken.  🤦

**Comment by jasongrout** [comment]
I thought I was too. Looks like @blink1073 and @minrk are owners of the `notebook` package on pypi, though - maybe they could make both of us maintainers too?

**Comment by blink1073** [comment]
I am not an owner, only a maintainer.

**Comment by Zsailer** [comment]
Notebook 6.1.6 released with this PR included.

**Pull request merged**
